gpytorch/lazy/diag_lazy_tensor.py

```diff
@@ -3,159 +3,131 @@
 from __future__ import print_function
 from __future__ import unicode_literals
 
+from itertools import product
 import torch
+from .added_diag_lazy_tensor import AddedDiagLazyTensor
 from .lazy_tensor import LazyTensor
+from ..utils.broadcasting import _mul_broadcast_shape
 
 
 class DiagLazyTensor(LazyTensor):
-    def __init__(self, diag, exact_root_decomposition=False):
+    def __init__(self, diag):
         """
-        Diagonal lazy tensor
+        Diagonal lazy tensor. Supports arbitrary batch sizes.
 
         Args:
-            - diag (tensor) (batch) diagonal of matrix
-            - exact_root_decomposition (bool, default False) return an exact root decomposition.
-                Set to True only if the represented diagonal is relatively small.
+            :attr:`diag` (Tensor):
+                A `b1 x ... x bk x n` Tensor, representing a `b1 x ... x bk`-sized batch
+                of `n x n` diagonal matrices
         """
         super(DiagLazyTensor, self).__init__(diag)
         self._diag = diag
-        self.exact_root_decomposition = exact_root_decomposition
+        self._batch_shape = diag.shape[:-1]
 
-    def _matmul(self, rhs):
-        if rhs.ndimension() == 1 and self.ndimension() == 2:
-            return self._diag * rhs
-        else:
-            res = self._diag.unsqueeze(-1).expand_as(rhs) * rhs
-            return res
+    def __add__(self, other):
+        if isinstance(other, DiagLazyTensor):
+            return self.add_diag(other._diag)
+        return AddedDiagLazyTensor(other, self)
 
-    def _t_matmul(self, rhs):
-        # Matrix is symmetric
-        return self._matmul(rhs)
+    def _get_indices(self, indices):
+        """Extract elements from the LazyTensor. Supports arbitrary batch sizes and broadcasting across batch sizes.
+
+        Args:
+            :attr:`indices` (LongTensor):
+                A `b1 x ... bk x 2 x j` LongTensor, for each (multi-dimensional) batch selecting j
+                elements from the associated `n x n` diagonal matrix. The `2 x j` tensor `t` associated
+                with the last two dimensions represents the indices as lists of row and column indices,
+                that is, selecting the `b1 x ... x bk x t(0, l) x t(1, l)`-th element.
+                The batch indices follow standard broadcasting rules.
+
+        """
+        batch_shape = _mul_broadcast_shape(indices.shape[:-2], self._diag.shape[:-1])
+        row_indices = indices[..., 0, :].expand(batch_shape + indices.shape[-1:])
+        diag = self._diag.expand(batch_shape + self._diag.shape[-1:])
+
+        diag_elements = torch.stack(
```

Comment:
yeah I know I said things would be easier... but supporting arbitrary batch shapes in both the indices as well as the underlying `DiagLazyTensor` as well as broadcasting across the batch shapes requires some ... finesse.

Reply:
Alternatively, we could do `(*batch_indices, row_indices, col_indices)`. And since this is an internal method, it makes sense that we you would strictly need to supply one index per tensor dimension.

Reply:
So the proposed interface supports selecting different elements for each batch, as well as broadcasting over arbitrary batch dimensions between the input indices and the actual batch size of the DiagLazyTensor. 

@gpleiss, sounds like you're suggesting to keep this simple and just get the same indices for each batch. In this case we can default to returning elements of all batches if no batch_indices are provided, and returning those of the specific batch if they are.

This would be quite different and less flexible than what I have currently, but on the other hand, it might be good not to overcomplicate things for now.

Reply:
I've been playing with this a bit, and in general, the `_batch_get_indices` and `_get_indices` methods have inconsistent behavior on master, especially compared to what happens with torch tensors. For example:

```python
import torch
from gpytorch.lazy import NonLazyTensor, DiagLazyTensor, ToeplitzLazyTensor
torch_mat = torch.randn(3, 5, 5)
non_lt = NonLazyTensor(torch_mat)
diag_lt = DiagLazyTensor(torch.randn(3, 5))
toep_lt = ToeplitzLazyTensor(torch.randn(3, 5))

print(torch_mat.shape, non_lt.shape, diag_lt.shape, toep_lt.shape) # All are torch.Size([3, 5, 5])

index = ([0, 1], [0, 0], [0, 1])
print(torch_mat[index].shape)  # torch.Size([2])
print(non_lt[index].shape)  # torch.Size([2])
print(diag_lt[index].shape)  # torch.Size([2, 2, 2])
print(toep_lt[index].shape)  # torch.Size([2, 2, 2])
```
(Also, `diag_lt.shape` returns a tuple instead of a `torch.Size`).

I think we should standardize these, and in my opinion the clearly correct choice is to mirror what happens with the torch tensor: `mat[[0, 0], [0, 0], [0, 1]]` should return a tensor containing `mat[0, 0, 0]` and `mat[0, 0, 1]`.

Reply:
Actually, my point appears to be a more minor bug dealing with lists vs tensors as indexes:
```python
print(toep_lt[[0, 1], [0, 0], [0, 1]].shape)  # torch.Size([2, 2, 2]) -- Bad!
a = torch.tensor([0, 1])
b = torch.tensor([0, 0])
print(toep_lt[a, b, a].shape)  # torch.Size([2]) -- Good!
```